tailnet/conn.go

```diff
@@ -569,54 +582,60 @@ func (c *Conn) BlockEndpoints() bool {
 	return c.blockEndpoints
 }
 
-// AwaitReachable pings the provided IP continually until the
-// address is reachable. It's the callers responsibility to provide
-// a timeout, otherwise this function will block forever.
-func (c *Conn) AwaitReachable(ctx context.Context, ip netip.Addr) bool {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel() // Cancel all pending pings on exit.
-
-	completedCtx, completed := context.WithCancel(context.Background())
-	defer completed()
-
-	run := func() {
-		// Safety timeout, initially we'll have around 10-20 goroutines
-		// running in parallel. The exponential backoff will converge
-		// around ~1 ping / 30s, this means we'll have around 10-20
-		// goroutines pending towards the end as well.
-		ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
-		defer cancel()
-
-		_, _, _, err := c.Ping(ctx, ip)
-		if err == nil {
-			completed()
-		}
-	}
-
-	eb := backoff.NewExponentialBackOff()
-	eb.MaxElapsedTime = 0
-	eb.InitialInterval = 50 * time.Millisecond
-	eb.MaxInterval = 30 * time.Second
-	// Consume the first interval since
-	// we'll fire off a ping immediately.
-	_ = eb.NextBackOff()
+var unixEpoch = time.Unix(0, 0)
 
-	t := backoff.NewTicker(eb)
-	defer t.Stop()
+// AwaitReachable waits until the peer with given IP is reachable, which we define as having handshaked at least once at
+// the Wireguard layer. It's the callers responsibility to provide  a timeout, otherwise this function will block
+// forever.
+func (c *Conn) AwaitReachable(ctx context.Context, ip netip.Addr) bool {
+	rightAway := make(chan struct{}, 1)
+	rightAway <- struct{}{}
+	tkr := time.NewTicker(50 * time.Millisecond)
+	defer tkr.Stop()
 
-	go run()
 	for {
 		select {
-		case <-completedCtx.Done():
-			return true
-		case <-t.C:
-			// Pings can take a while, so we can run multiple
-			// in parallel to return ASAP.
-			go run()
 		case <-ctx.Done():
 			return false
+		case <-tkr.C:
+		case <-rightAway:
+		}
+
+		nKeys := c.getNodeKeysForIP(ip)
+		if len(nKeys) == 0 {
+			c.logger.Debug(ctx, "missing node(s) for IP", slog.F("ip", ip.String()))
+			continue
+		}
+		s := c.Status()
```

Comment:
I'm considering this out of the "single tailnet" perspective, should we instead rely on utilizing the peers reported in the status callback (`SetStatusCallback`)? Let's say this is the server, and there are a lot of peers and traffic, then waiting for multiple connections will result in a lot of wg state reporting/mutex locking along the way.

This method could also let us signal this loop when a peer is available instead of polling it on an arbitrary timer.

Just a thought.

Reply:
I like where you're head is at.

I did think about this, but decided not to build it until/unless we get some evidence it's a problem at scale.  The code I'm replacing had us sending a bunch of pings, and who knows how expensive that was!  If we see extended lag when making a lot of coderd proxy'd connections at scale, we can look into improving the efficiency here.